db/view/view.cc

```diff
@@ -2623,7 +2626,137 @@ future<> view_builder::migrate_to_v2(locator::token_metadata_ptr tmptr, db::syst
+                slice,
+                std::move(trace_state),
+                streamed_mutation::forwarding::no,
+                fwd_mr);
+    }
+    virtual const dht::partition_range* get_read_range() const override {
+        const auto shard = this_shard_id();
+        return _contexts[shard].range.get();
+    }
+    virtual void update_read_range(lw_shared_ptr<const dht::partition_range> range) override {
+        const auto shard = this_shard_id();
+        _contexts[shard].range = make_foreign(std::move(range));
+    }
+    virtual future<> destroy_reader(stopped_reader reader) noexcept override {
+        auto ctx = std::move(_contexts[this_shard_id()]);
+        auto reader_opt = ctx.semaphore->unregister_inactive_read(std::move(reader.handle));
+        if  (!reader_opt) {
+            return make_ready_future<>();
+        }
+        return reader_opt->close().finally([ctx = std::move(ctx)] {});
+    }
+    virtual reader_concurrency_semaphore& semaphore() override {
+        const auto shard = this_shard_id();
+        if (!_contexts[shard].semaphore) {
+            _contexts[shard].semaphore = &_db.local().get_reader_concurrency_semaphore();
+        }
+        return *_contexts[shard].semaphore;
+    }
+    virtual future<reader_permit> obtain_reader_permit(schema_ptr schema, const char* const description, db::timeout_clock::time_point timeout, tracing::trace_state_ptr trace_ptr) override {
+        return make_ready_future<reader_permit>(semaphore().make_tracking_only_permit(schema, description, timeout, std::move(trace_ptr)));
+    }
+};
+
+void view_builder::init_virtual_table() {
+    if (_view_build_status_on_group0) {
+        // We set the old system_distributed.view_build_status table to read virtually
+        // from system.view_build_status_v2 in order to make the transition transparent for
+        // readers of the table and maintain compatibility.
+
+        auto ms_v2 = mutation_source([this] (schema_ptr s,
+                reader_permit permit,
+                const dht::partition_range& pr,
+                const query::partition_slice& ps,
+                tracing::trace_state_ptr trace_state,
+                streamed_mutation::forwarding,
+                mutation_reader::forwarding fwd_mr) {
+
+            // The v1 distributed table and the v2 system local table have different shard mapping.
+            // The v2 table uses the null sharder, everything is on shard zero, while the v1 table
+            // has the normal sharder.
+            // So to simulate a read in v1 table in some shard, we need to read all the rows
+            // belonging to this shard from v2 table in shard zero.
+            // In order to read the table in a different shard we use the multishard reader. It is
+            // set to read from all the shards, but actually it only has rows to read in shard zero.
+            // Then we also need to filter only the keys belonging to this shard according to v1 sharder.
```

Comment:
You can get away without filtering, if you fast forward between the range. That said both filtering and fast-forwarding has their pros/cons and I think filtering will work better for small amount of data, which this table should have.

Reply:
I considered also to split the range into ranges owned by this shard, but the solution with filtering was simpler for me.
Also, I expect there are not many reads from this table, and if there are they can be changed to read from the v2 table instead.